From acccc93e084641861ca553317edb7da7791833b5 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@openwrt.org>
Date: Tue, 21 Oct 2014 22:24:50 +0200
Subject: [PATCH] getopt: fix optional argument processing

Processing an option character with optional argument fails if the
option is last on the command line. This happens because the
if (optind >= argc) check runs first before testing for optional
argument.
---
 src/misc/getopt.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/misc/getopt.c b/src/misc/getopt.c
index 8a2e4d5..f94c4f7 100644
--- a/src/misc/getopt.c
+++ b/src/misc/getopt.c
@@ -55,7 +55,8 @@ int getopt(int argc, char * const argv[], const char *optstring)
 		return '?';
 	}
 	if (optstring[i+1] == ':') {
-		if (optind >= argc) {
+		if (optstring[i+2] == ':') optarg = 0;
+		else if (optind >= argc) {
 			if (optstring[0] == ':') return ':';
 			if (opterr) {
 				write(2, argv[0], strlen(argv[0]));
@@ -65,7 +66,6 @@ int getopt(int argc, char * const argv[], const char *optstring)
 			}
 			return '?';
 		}
-		if (optstring[i+2] == ':') optarg = 0;
 		if (optstring[i+2] != ':' || optpos) {
 			optarg = argv[optind++] + optpos;
 			optpos = 0;
-- 
2.2.0

